silence/silence.go

```diff
@@ -544,104 +540,179 @@ func validateSilence(s *pb.Silence) error {
+	if expireExisting {
+		if err := s.expire(existing.Id); err != nil {
+			return fmt.Errorf("failed to expire silence: %w", err)
+		}
 	}
 
-	// Check the limit unless the silence has been expired. This is to avoid
-	// situations where silences cannot be expired after the limit has been
-	// reduced.
+	return s.mergeAndBroadcast(msil, now)
+}
+
+func (s *Silences) checkLimits(msil *pb.MeshSilence, updateExisting bool) error {
 	if s.limits.MaxSilenceSizeBytes != nil {
 		n := msil.Size()
-		if m := s.limits.MaxSilenceSizeBytes(); m > 0 && n > m && sil.EndsAt.After(now) {
-			return fmt.Errorf("silence exceeded maximum size: %d bytes (limit: %d bytes)", n, m)
+		if m := s.limits.MaxSilenceSizeBytes(); m > 0 && n > m {
+			return fmt.Errorf("maximum size: %d bytes (limit: %d bytes)", n, m)
+		}
+	}
+	if s.limits.MaxSilences != nil && !updateExisting {
+		if m := s.limits.MaxSilences(); m > 0 && len(s.st)+1 > m {
+			return fmt.Errorf("maximum number of silences: %d (limit: %d)", len(s.st), m)
 		}
 	}
+	return nil
+}
 
+func (s *Silences) getSilence(id string) (*pb.Silence, bool) {
+	msil, ok := s.st[id]
+	if !ok {
+		return nil, false
+	}
+	return msil.Silence, true
+}
+
+func (s *Silences) mergeAndBroadcast(msil *pb.MeshSilence, now time.Time) error {
+	b, err := marshalMeshSilence(msil)
+	if err != nil {
+		return fmt.Errorf("failed to marshal silence: %w", err)
+	}
 	if s.st.merge(msil, now) {
 		s.version++
 	}
 	s.broadcast(b)
-
 	return nil
 }
 
-// Set the specified silence. If a silence with the ID already exists and the modification
-// modifies history, the old silence gets expired and a new one is created.
-func (s *Silences) Set(sil *pb.Silence) error {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
+func (s *Silences) toMeshSilence(sil *pb.Silence) *pb.MeshSilence {
+	return &pb.MeshSilence{
+		Silence:   sil,
+		ExpiresAt: sil.EndsAt.Add(s.retention),
+	}
+}
 
-	now := s.nowUTC()
-	prev, ok := s.getSilence(sil.Id)
-	if sil.Id != "" && !ok {
+// Expire the silence with the given ID immediately.
+// It is idempotent, nil is returned if the silence already expired before it is GC'd.
+// If the silence is not found an error is returned.
+func (s *Silences) expire(id string) error {
+	sil, ok := s.getSilence(id)
+	if !ok {
 		return ErrNotFound
 	}
+	sil = shallowClone(sil)
+	now := s.nowUTC()
 
-	if ok {
-		if canUpdate(prev, sil, now) {
-			return s.setSilence(sil, now, false)
-		}
-		if getState(prev, s.nowUTC()) != types.SilenceStateExpired {
-			// We cannot update the silence, expire the old one.
-			if err := s.expire(prev.Id); err != nil {
-				return fmt.Errorf("expire previous silence: %w", err)
-			}
-		}
+	switch getState(sil, now) {
+	case types.SilenceStateExpired:
+		return nil
+	case types.SilenceStateActive:
+		sil.EndsAt = now
+	case types.SilenceStatePending:
+		// Set both to now to make Silence move to "expired" state
+		sil.StartsAt = now
+		sil.EndsAt = now
 	}
+	sil.UpdatedAt = now
```

Comment:
Since `UpdatedAt` is no longer updated in `silenceSilence`, it must be updated here too. This is covered by a number of existing unit tests:

```
--- FAIL: TestSilenceSet (0.00s)
    silence_test.go:451:
        	Error Trace:	/Users/grobinson/go/src/github.com/prometheus/alertmanager/silence/silence_test.go:451
        	Error:      	Not equal:
        	            	expected: silence.state{"019d01ce-56a9-45f3-86ae-f7b367f6e0a2":(*silencepb.MeshSilence)(0x14000253580), "55de6030-4cad-4a54-9469-47b29d49dbda":(*silencepb.MeshSilence)(0x14000253340), "6e844958-ae9d-4288-b23c-87f184c9615b":(*silencepb.MeshSilence)(0x140002535c0)}
        	            	actual  : silence.state{"019d01ce-56a9-45f3-86ae-f7b367f6e0a2":(*silencepb.MeshSilence)(0x14000253480), "55de6030-4cad-4a54-9469-47b29d49dbda":(*silencepb.MeshSilence)(0x14000253300), "6e844958-ae9d-4288-b23c-87f184c9615b":(*silencepb.MeshSilence)(0x14000253500)}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -22,3 +22,3 @@
        	            	     wall: (uint64) 0,
        	            	-    ext: (int64) 62135596980,
        	            	+    ext: (int64) 62135602920,
        	            	     loc: (*time.Location)(<nil>)
        	            	@@ -27,3 +27,3 @@
        	            	     wall: (uint64) 0,
        	            	-    ext: (int64) 62135596980,
        	            	+    ext: (int64) 62135596920,
        	            	     loc: (*time.Location)(<nil>)
        	            	@@ -40,3 +40,3 @@
        	            	    wall: (uint64) 0,
        	            	-   ext: (int64) 62135600580,
        	            	+   ext: (int64) 62135606520,
        	            	    loc: (*time.Location)(<nil>)
        	Test:       	TestSilenceSet
        	Messages:   	unexpected state after silence creation
--- FAIL: TestLimits (0.00s)
    silence_test.go:511:
        	Error Trace:	/Users/grobinson/go/src/github.com/prometheus/alertmanager/silence/silence_test.go:511
        	Error:      	Not equal:
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestLimits
--- FAIL: TestSilenceExpire (0.00s)
    silence_test.go:1056:
        	Error Trace:	/Users/grobinson/go/src/github.com/prometheus/alertmanager/silence/silence_test.go:1056
        	Error:      	Not equal:
        	            	expected: &silencepb.Silence{Id:"pending", Matchers:[]*silencepb.Matcher{(*silencepb.Matcher)(0x14000257630)}, StartsAt:time.Date(1970, time.January, 1, 0, 0, 0, 0, time.UTC), EndsAt:time.Date(1970, time.January, 1, 0, 0, 0, 0, time.UTC), UpdatedAt:time.Date(1970, time.January, 1, 0, 0, 0, 0, time.UTC), Comments:[]*silencepb.Comment(nil), CreatedBy:"", Comment:"", XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
        	            	actual  : &silencepb.Silence{Id:"pending", Matchers:[]*silencepb.Matcher{(*silencepb.Matcher)(0x14000257630)}, StartsAt:time.Date(1970, time.January, 1, 0, 1, 0, 0, time.UTC), EndsAt:time.Date(1970, time.January, 1, 1, 0, 0, 0, time.UTC), UpdatedAt:time.Date(1969, time.December, 31, 23, 0, 0, 0, time.UTC), Comments:[]*silencepb.Comment(nil), CreatedBy:"", Comment:"", XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -15,3 +15,3 @@
        	            	   wall: (uint64) 0,
        	            	-  ext: (int64) 62135596800,
        	            	+  ext: (int64) 62135596860,
        	            	   loc: (*time.Location)(<nil>)
        	            	@@ -20,3 +20,3 @@
        	            	   wall: (uint64) 0,
        	            	-  ext: (int64) 62135596800,
        	            	+  ext: (int64) 62135600400,
        	            	   loc: (*time.Location)(<nil>)
        	            	@@ -25,3 +25,3 @@
        	            	   wall: (uint64) 0,
        	            	-  ext: (int64) 62135596800,
        	            	+  ext: (int64) 62135593200,
        	            	   loc: (*time.Location)(<nil>)
        	Test:       	TestSilenceExpire
--- FAIL: TestSilenceExpireWithZeroRetention (0.00s)
    silence_test.go:1167:
        	Error Trace:	/Users/grobinson/go/src/github.com/prometheus/alertmanager/silence/silence_test.go:1167
        	Error:      	Not equal:
        	            	expected: 0
        	            	actual  : 1
        	Test:       	TestSilenceExpireWithZeroRetention
--- FAIL: TestSilenceExpireInvalid (0.00s)
    silence_test.go:1212:
        	Error Trace:	/Users/grobinson/go/src/github.com/prometheus/alertmanager/silence/silence_test.go:1212
        	Error:      	Not equal:
        	            	expected: 0
        	            	actual  : 1
        	Test:       	TestSilenceExpireInvalid
FAIL
FAIL	github.com/prometheus/alertmanager/silence	0.249s
FAIL
```